### ISSUE TITLE
Implement Teuchos MPI operations

### ISSUE THREAD

**Issue by cwpearson** [kokkos/kokkos-comm#10]
The motivation here is to prove we have a reasonable implementation by being able to run Trilinos on top of KokkosComm

| MPI_... | 1D | non-contiguous |
|-|-|-|
| allreduce | |
| broadcast | |
| gather | |
| scan | |

- [ ] abort
- [x] allgather
- [x] barrier
- [x] irecv
  - [ ] non-contiguous [\#32](https://github.com/kokkos/kokkos-comm/pull/32)
- [x] rsend
- [x] recv
- [x] send
- [x] ssend

**Comment by cedricchevalier19** [comment]
Very good starting point.

I have a comment concerning the API. Here we have 3 "blocking" `send` and only one "immediate". 
It might make sense to have only one of each and a template parameter to choose between the implementation.

```C++
send(snd_buf, comm); // Alias for send<Default>(snd_buff, comm);
send<Ready>(snd_buf, comm);
send<Synchronous>(snd_buf, comm);
send<Buffered>(snd_buf, comm);
```

This way we can also implement the `isend` variants. I think we should try to push users to `I` variants as they are a lot easier to use to make a semantically sound code, and also be ready for more asynchronous computations.

**Comment by cwpearson** [comment]
That's definitely an option, but since the `Immediate` version returns `KokkosComm::Req` and the rest do not, you actually have to do something like:

```c++
enum class Mode {
    Blocking,
    Buffered,
    Immediate,
    Ready,
    Synchronous
};

// most modes just return an error code
template <Mode mode>
struct SendReturn {
    using type = int;
}

// Immediate return value includes an MPI_Request
struct IsendResult {
    int err;
    MPI_Request req;
};
template<> struct SendReturn<Mode::Immediate> {
    using type = IsendResult;
}

template <Mode mode>
typename SendReturn<mode>::type send(...);

template<> typename SendReturn<Blocking>::type send<Blocking>(...) {...}
...
```

I think (though I might be missing something), this basically just replaces symbol lookup with template resolution in the compiler, and I'm not sure it really buys us anything

**Comment by cwpearson** [comment]
I see that I misunderstood a bit - you were just suggesting doing this for the synchronous versions. I think my point still stands (though we can avoid the funky return type template).

**Comment by cedricchevalier19** [comment]
I think `send` and `isend` should be separate functions as they are semantically very different.
But between variants they are quite interchangeable.

The idea of the template parameter with a default value is to be able to change this value for ease of checking communication pattern correctness, like force synchronous send instead of send to prevent deadlocks that can be hidden by a high eager limit on the test platform.

So basically, in pseudo-code
```C++
enum class Mode {
    Bare,
    Blocking,
    Ready,
    Synchronous
};

#ifdef DEBUG_COMM
template<Mode mode=Synchronous>
#else
template<Mode mode=Bare>
#endif
int send(View, Comm);

#ifdef DEBUG_COMM
template<Mode mode=Synchronous>
#else
template<Mode mode=Bare>
#endif
Req isend(View, Comm);
```

Or we can go crazier and have a `CommPolicy` object to set the default.

**Comment by cwpearson** [comment]
Maybe the thing to do right now while we consider this is put some relatively lean wrappers around MPI functions in the `Impl` namespace, since we'll almost certainly need those anyway.

**Comment by nicoleavans** [comment]
I surveyed what is used in Tpetra and Teuchos/Comm. It is broken down into what is needed for both, and each individually. 
## Overall
❌Overall Missing:
- [ ] `MPI_Alltoallv` 
- [ ] `MPI_Cancel` 
- [ ] `MPI_Gather` 
- [ ] `MPI_Gatherv` 
- [ ] `MPI_Iallreduce` 
- [ ] `MPI_Scan` (See PR #168)
- [ ] `MPI_Scatter` 
- [ ] `MPI_Scatterv` 

❓Do we need to wrap? (I'm thinking no, *particularly* the `MPI_Comm_` related features. They are included for completeness and discussion.) 
- [ ] `MPI_Abort` 
- [ ] `MPI_Comm` 
	- [ ] `MPI_Comm_Compare`
	- [ ] `MPI_Comm_create` 
	- [ ] `MPI_Comm_create_keyval` 
	- [ ] `MPI_Comm_dup` 
	- [ ] `MPI_Comm_free` 
	- [ ] `MPI_Comm_free_keyval` 
	- [ ] `MPI_Comm_group` 
	- [ ] `MPI_Comm_set_errhandler` 
	- [ ] `MPI_Comm_size` 
	- [ ] `MPI_Comm_split` 
	- [ ] `MPI_Comm_rank` 
	- [ ] `MPI_Comm_test_inter` 
- [ ] `MPI_Errhandler_set` 
- [ ] `MPI_Error_string` 
- [ ] `MPI_Finalized` 
- [ ] `MPI_Group_free` 
- [ ] `MPI_Group_incl` 
- [ ] `MPI_Initialized` 
- [ ] `MPI_Type_commit` 
- [ ] `MPI_Type_contiguous` 
- [ ] `MPI_Type_create_struct` 
- [ ] `MPI_Type_free` 
- [ ] `MPI_Type_size` 
- [ ] `MPI_Type_struct` 

✅Overall Implemented: 
- [x] `MPI_Allgather` 
- [x] `MPI_Allreduce` 
- [x] `MPI_Barrier` 
- [x] `MPI_Bcast` 
- [x] `MPI_Irecv` 
- [x] `MPI_Isend` 
- [x] `MPI_Recv` 
- [x] `MPI_Reduce` 
- [x] `MPI_Send` 
- [x] `MPI_Ssend` 
- [x] `MPI_Wait` 
- [x] `MPI_Waitall` (essentially [implemented](https://github.com/kokkos/kokkos-comm/blob/8f15a9066bef1bed9d2da5b3f8866399219c26e9/src/KokkosComm/mpi/req.hpp#L80))
## Tpetra
❌Tpetra Missing:
- [ ] `MPI_Alltoallv` 
- [ ] `MPI_Gather` 
- [ ] `MPI_Gatherv` 
- [ ] `MPI_Iallreduce` 

❓Do we need to wrap? (I'm thinking no...) 
- [ ] `MPI_Abort` 
- [ ] `MPI_Comm_Compare`
- [ ] `MPI_Comm_test_inter` 
- [ ] `MPI_Finalized` 
- [ ] `MPI_Initialized` 
- [ ] `MPI_Type_commit` 
- [ ] `MPI_Type_contiguous` 
- [ ] `MPI_Type_create_struct` 
- [ ] `MPI_Type_struct` 

✅Tpetra Implemented: 
- [x] `MPI_Allgather` 
- [x] `MPI_Allreduce` 
- [x] `MPI_Irecv` 
- [x] `MPI_Isend` 
- [x] `MPI_Waitall`  (essentially [implemented](https://github.com/kokkos/kokkos-comm/blob/8f15a9066bef1bed9d2da5b3f8866399219c26e9/src/KokkosComm/mpi/req.hpp#L80))

- [Tpetra_ComputeGatherMap.hpp](https://github.com/trilinos/Trilinos/blob/master/packages/tpetra/core/src/Tpetra_ComputeGatherMap.hpp)
	- [ ] `MPI_Gather` ❌
	- [ ] `MPI_Gatherv` ❌
- [Tpetra_Details_checkGlobalError.cpp](https://github.com/trilinos/Trilinos/blob/master/packages/tpetra/core/src/Tpetra_Details_checkGlobalError.cpp)
	- [ ] `MPI_Abort` ❓
- [Tpetra_Details_DistributorActor.hpp](https://github.com/trilinos/Trilinos/blob/master/packages/tpetra/core/src/Tpetra_Details_DistributorActor.hpp)
	- [ ] `MPI_Alltoallv` ❌
- [Tpetra_Details_DistributorPlan.cpp](https://github.com/trilinos/Trilinos/blob/master/packages/tpetra/core/src/Tpetra_Details_DistributorPlan.cpp)
	- [ ] `MPIX_Dist_graph_create_adjacent` ❌
		- [MPIX Implementation](https://github.com/mpi-advance/locality_aware/blob/e09eb2ac9dab6a1cc2bd8f6f8e5d71ad64e9f152/src/neighborhood/dist_graph.c)
		- [MPI Standard](https://www.mpi-forum.org/docs/mpi-5.0/mpi50-report.pdf) Page 386
			- [OpenMPI](https://www.open-mpi.org/doc/v3.0/man3/MPI_Dist_graph_create_adjacent.3.php)
			- [MPICH](https://www.mpich.org/static/docs/v3.3/www3/MPI_Dist_graph_create_adjacent.html)
- [Tpetra_Details_iallreduce.cpp](https://github.com/trilinos/Trilinos/blob/master/packages/tpetra/core/src/Tpetra_Details_iallreduce.cpp)
	- [ ] `MPI_Iallreduce` ❌
	- [x] `MPI_Allreduce` 
- [Tpetra_Details_isInterComm.cpp](https://github.com/trilinos/Trilinos/blob/master/packages/tpetra/core/src/Tpetra_Details_isInterComm.cpp)
	- [ ] `MPI_Comm_test_inter` ❓
- [Tpetra_Details_mpiIsInitialized.cpp](https://github.com/trilinos/Trilinos/blob/master/packages/tpetra/core/src/Tpetra_Details_mpiIsInitialized.cpp)
	- [ ] `MPI_Initialized` ❓
	- [ ] `MPI_Finalized` ❓
- [Tpetra_Details_MpiTypeTraits.cpp](https://github.com/trilinos/Trilinos/blob/master/packages/tpetra/core/src/Tpetra_Details_MpiTypeTraits.cpp)
	- [ ] `MPI_Type_contiguous` ❓
	- [ ] `MPI_Type_struct` ❓
	- [ ] `MPI_Type_create_struct` ❓
	- [ ] `MPI_Type_commit` ❓
- [Tpetra_Details_printOnce.cpp](https://github.com/trilinos/Trilinos/blob/master/packages/tpetra/core/src/Tpetra_Details_printOnce.cpp)
	- [ ] `MPI_Initialized` ❓
	- [ ] `MPI_Finalized` ❓
- [Tpetra_DirectoryImpl_def.hpp](https://github.com/trilinos/Trilinos/blob/master/packages/tpetra/core/src/Tpetra_DirectoryImpl_def.hpp)
	- [x] `MPI_Allgather` 
- [Tpetra_Import_Util2.hpp](https://github.com/trilinos/Trilinos/blob/master/packages/tpetra/core/src/Tpetra_Import_Util2.hpp)
	- [x] `MPI_Irecv` 
	- [x] `MPI_Waitall` 
	- [x] `MPI_Isend` 
	- [ ] `MPI_Abort` ❓
- [Tpetra_Util.cpp](https://github.com/trilinos/Trilinos/blob/master/packages/tpetra/core/src/Tpetra_Util.cpp)
	- `MPI_Comm_Compare`❓

## Teuchos Comm
❌Teuchos Comm Missing:
- [ ] `MPI_Cancel` 
- [ ] `MPI_Gather` 
- [ ] `MPI_Gatherv` 
- [ ] `MPI_Scan` 
- [ ] `MPI_Scatter` 
- [ ] `MPI_Scatterv` 

❓Do we need to wrap? (I'm thinking no...)
- [ ] `MPI_Comm_dup` 
- [ ] `MPI_Comm_create` 
- [ ] `MPI_Comm_create_keyval` 
- [ ] `MPI_Comm_free` 
- [ ] `MPI_Comm_free_keyval` 
- [ ] `MPI_Comm_group` 
- [ ] `MPI_Comm_set_errhandler` 
- [ ] `MPI_Comm_size` 
- [ ] `MPI_Comm_split` 
- [ ] `MPI_Errhandler_set` 
- [ ] `MPI_Error_string` 
- [ ] `MPI_Finalized` 
- [ ] `MPI_Group_free` 
- [ ] `MPI_Group_incl` 
- [ ] `MPI_Initialized` 
- [ ] `MPI_Type_commit` 
- [ ] `MPI_Type_contiguous` 
- [ ] `MPI_Type_create_struct` 
- [ ] `MPI_Type_free` 
- [ ] `MPI_Type_size` 
- [ ] `MPI_Type_struct` 

✅Teuchos Comm Implemented: 
- [x] `MPI_Allgather` 
- [x] `MPI_Allreduce` 
- [x] `MPI_Barrier` 
- [x] `MPI_Bcast` 
- [x] `MPI_Irecv` 
- [x] `MPI_Isend` 
- [x] `MPI_Recv` 
- [x] `MPI_Reduce` 
- [x] `MPI_Rsend` 
- [x] `MPI_Send` 
- [x] `MPI_Ssend` 
- [x] `MPI_Waitall` (essentially [implemented](https://github.com/kokkos/kokkos-comm/blob/8f15a9066bef1bed9d2da5b3f8866399219c26e9/src/KokkosComm/mpi/req.hpp#L80))

- [teuchos/comm/src/Teuchos_CommHelpers.cpp](https://github.com/trilinos/Trilinos/blob/master/packages/teuchos/comm/src/Teuchos_CommHelpers.cpp)
	- [ ] `MPI_Error_string` ❓
	- [x] `MPI_Allreduce` 
	- [ ] `MPI_Gather`❌
	- [ ] `MPI_Scatter` ❌
	- [ ] `MPI_Scatterv` ❌
	- [x] `MPI_Reduce` 
	- [ ] `MPI_Gatherv` ❌
	- [x] `MPI_Irecv` 
	- [x] `MPI_Send` 
	- [x] `MPI_Isend` 
	- [x] `MPI_Wait` 
- [teuchos/comm/src/Teuchos_DefaultComm.hpp](https://github.com/trilinos/Trilinos/blob/master/packages/teuchos/comm/src/Teuchos_DefaultComm.hpp)
	- [ ] `MPI_Comm_create_keyval` ❓
	- [ ] `MPI_Comm_set_attr` ❓
	- [ ] `MPI_Comm_free_keyval` ❓
- [teuchos/comm/src/Teuchos_DefaultMpiComm.cpp](https://github.com/trilinos/Trilinos/blob/master/packages/teuchos/comm/src/Teuchos_DefaultMpiComm.cpp)
	- [ ] `MPI_Error_string` ❓
	- [ ] `MPI_Finalized` ❓
	- [ ] `MPI_Comm_free` ❓
	- [ ] `MPI_Comm_set_errhandler` ❓
	- [ ] `MPI_Errhandler_set` ❓
	- [ ] `MPI_Error_string` ❓
- [teuchos/comm/src/Teuchos_DefaultMpiComm.hpp](https://github.com/trilinos/Trilinos/blob/master/packages/teuchos/comm/src/Teuchos_DefaultMpiComm.hpp)
	- [x] `MPI_Wait` 
	- [ ] `MPI_Cancel` ❌
	- [ ] `MPI_Comm_size` ❓
	- [ ] `MPI_Comm_rank` ❓
	- [ ] `MPI_Comm_dup` ❓
	- [x] `MPI_Bcast` 
	- [x] `MPI_Barrier` 
	- [x] `MPI_Allgather` 
	- [ ] `MPI_Gather` ❌
	- [ ] `MPI_Type_contiguous` ❓
	- [ ] `MPI_Type_commit` ❓
	- [x] `MPI_Allreduce` 
	- [ ] `MPI_Type_free` ❓
	- [ ] `MPI_Scan` ❌
	- [x] `MPI_Send` 
	- [x] `MPI_Ssend` 
	- [x] `MPI_Rsend` 
	- [x] `MPI_Recv` 
	- [x] `MPI_Isend` 
	- [x] `MPI_Irecv` 
	- [x] `MPI_Waitall` 
	- [ ] `MPI_Comm_split` ❓
	- [ ] `MPI_Comm_group` ❓
	- [ ] `MPI_Group_free` ❓
	- [ ] `MPI_Group_incl` ❓
	- [ ] `MPI_Comm_create` ❓
- [teuchos/comm/src/Teuchos_Details_MpiTypeTraits.hpp](https://github.com/trilinos/Trilinos/blob/master/packages/teuchos/comm/src/Teuchos_Details_MpiTypeTraits.hpp)
	- [ ] `MPI_Type_struct` ❓
	- [ ] `MPI_Type_create_struct` ❓
	- [ ] `MPI_Type_commit` ❓
- [teuchos/comm/src/Teuchos_MpiReductionOpSetter.cpp](https://github.com/trilinos/Trilinos/blob/master/packages/teuchos/comm/src/Teuchos_MpiReductionOpSetter.cpp)
	- [ ] `MPI_Comm_create_keyval` ❓
	- [ ] `MPI_Comm_free_keyval` ❓
- [teuchos/comm/src/Teuchos_MpiReductionOpSetter.hpp](https://github.com/trilinos/Trilinos/blob/master/packages/teuchos/comm/src/Teuchos_MpiReductionOpSetter.hpp)
	- [ ] `MPI_Type_size` ❓
- [teuchos/comm/src/Teuchos_TimeMonitor.cpp](https://github.com/trilinos/Trilinos/blob/master/packages/teuchos/comm/src/Teuchos_TableFormat.cpp)
	- [ ] `MPI_Initialized` ❓

> `MPI_Request` objects are wrapped in [`req.hpp`](https://github.com/kokkos/kokkos-comm/blob/8f15a9066bef1bed9d2da5b3f8866399219c26e9/src/KokkosComm/mpi/req.hpp#L31), but are not included in this listing.
